src/main/java/fr/openent/competences/service/impl/DefaultExportService.java

```diff
@@ -1140,6 +1141,112 @@ && isInPeriode(devoirs.get(devoir).getString("date"), beginningYear, endingYear)
+                }
+            }));
+        }
+
+        Map<String, List<String>> devoirByCompetences = new HashMap<>();
+        for (JsonObject competence : competences.values()) {
+            competencesObjByIdComp.put(String.valueOf(competence.getLong(Field.ID_COMPETENCE)), competence);
+
+            if (competence.containsKey(Field.EMPTY)) {
+                continue;
+            }
+            String idDevoir = String.valueOf(competence.getLong(Field.ID_DEVOIR));
+            String idComp = String.valueOf(competence.getLong(Field.ID_COMPETENCE));
+            if (!devoirByCompetences.containsKey(idComp)) {
+                devoirByCompetences.put(idComp, new ArrayList<String>());
+            }
+            devoirByCompetences.get(idComp).add(idDevoir);
+            if (null != competence.getString(Field.ID_MATIERE)
+                    && null != competencesByMatiere.get(competence.getString(Field.ID_MATIERE).toString())) {
+                competencesByMatiere.get(competence.getString(Field.ID_MATIERE).toString()).add(idComp);
+            }
+        }
+
+        JsonObject bodyHeader = new JsonObject();
+        bodyHeader.put(Field.LEFT, "Matieres / items");
+        String right = getLibelle("evaluations.competence.level.and.number");
+        bodyHeader.put(Field.RIGHT, right);
+        body.put(Field.HEADER, bodyHeader);
+
+        for (Map.Entry<String, Set<String>> competencesInDomain : competencesByMatiere.entrySet()) {
+            JsonObject domainObj = new JsonObject();
+            if (matieres.get(competencesInDomain.getKey()) != null) {
+                domainObj.put(Field.DOMAINHEADER, matieres.get(competencesInDomain.getKey())
+                        .getString(Field.NAME));
+            }
+
+            JsonArray competencesInDomainArray = new fr.wseduc.webutils.collections.JsonArray();
+            List<Long> valuesByComp = new ArrayList<>();
+            for (String competence : competencesInDomain.getValue()) {
+                for (String devoir : devoirByCompetences.get(competence)) {
+                    if (competenceNotesByDevoir.containsKey(devoir) && competenceNotesByDevoir.get(devoir)
+                            .containsKey(competence)) {
+                        valuesByComp.add(competenceNotesByDevoir.get(devoir).get(competence) + 1);
+                    } else {
+                        valuesByComp.add(0L);
+                    }
+                }
+            }
+            JsonObject competenceNote = new JsonObject();
+            competenceNote.put(Field.HEADER, competencesInDomain.getKey());
+            competenceNote.put(Field.COMPETENCENOTES, calcWidthNote(text, usePerso, maitrises, valuesByComp, devoirs.size()));
+            competencesInDomainArray.add(competenceNote);
+            competencesArray.add(competenceNote);
+        }
+
+        body.put(Field.BODY, competencesArray);
+
+        result.put(Field.BODY, body);
+        return result;
+    }
```

Comment:
Plutôt longue cette fonction.
essaie de découper en sous méthodes (une manière simple de le faire: chaque fois que tu fais un for, ou une boucle, pose toi la question de ce qu'elle fait => et là tu as le nom de ta nouvelle méthode 😉).